cardano-cli/src/Cardano/CLI/Run/Friendly.hs

```diff
@@ -11,167 +11,203 @@
-    , "certificates" .= fmap textShow (Shelley._certs body)
-    , "withdrawals" .= Shelley.unWdrl (Shelley._wdrls body)
-    , "fee" .= Shelley._txfee body
-    , "time to live" .= Shelley._ttl body
-    , "update" .= fmap textShow (Shelley._txUpdate body)
-    , "metadata hash" .= fmap textShow (Shelley._mdHash body)
-    ]
-
-friendlyTxBodyAllegra
-  :: ShelleyMA.TxBody (ShelleyMAEra 'Allegra StandardCrypto) -> Object
-friendlyTxBodyAllegra
-  (ShelleyMA.TxBody
-    inputs
-    outputs
-    certificates
-    (Shelley.Wdrl withdrawals)
-    txfee
-    validity
-    update
-    adHash
-    _mint -- mint is not used in Allegra, only in Mary+
-    ) =
-  HashMap.fromList
-    [ "inputs" .= inputs
-    , "outputs" .= fmap friendlyTxOutAllegra outputs
-    , "certificates" .= fmap textShow certificates
-    , "withdrawals" .= withdrawals
-    , "fee" .= txfee
-    , "validity interval" .= friendlyValidityInterval validity
-    , "update" .= fmap textShow update
-    , "auxiliary data hash" .= fmap textShow adHash
+friendlyTxBodyBS
+  :: IsCardanoEra era => Api.TxBody era -> Either Prelude.String ByteString
+friendlyTxBodyBS =
+  fmap (encodePretty $ setConfCompare compare defConfig) . friendlyTxBody
+
+friendlyTxBody
+  :: forall era
+  .  IsCardanoEra era => Api.TxBody era -> Either Prelude.String Value
+friendlyTxBody txbody =
+  case getTransactionBodyContent txbody of
+    Right
+      TxBodyContent
+        { txIns
+        , txOuts
+        , txFee
+        , txValidityRange
+        , txMetadata
+        , txAuxScripts
+        , txWithdrawals
+        , txCertificates
+        , txUpdateProposal
+        , txMintValue
+        } ->
+      Right $
+      object
+        $   [ "era"     .= era
+            , "fee"     .= friendlyFee txFee
+            , "inputs"  .= friendlyInputs txIns
+            , "outputs" .= map friendlyTxOut txOuts
+            ]
+        ++  [ "auxiliary scripts" .= friendlyAuxScripts txAuxScripts
+            | Just _ <- [auxScriptsSupportedInEra era]
+            ]
+        ++  [ "certificates" .= friendlyCertificates txCertificates
+            | Just _ <- [certificatesSupportedInEra era]
+            ]
+        ++  [ "metadata" .= friendlyMetadata txMetadata
+            | Just _ <- [txMetadataSupportedInEra era]
+            ]
+        ++  [ "mint" .= friendlyMintValue txMintValue
+            | Right _ <- [multiAssetSupportedInEra era]
+            ]
+        ++  [ "update proposal" .= friendlyUpdateProposal txUpdateProposal
+            | Just _ <- [updateProposalSupportedInEra era]
+            ]
+        ++  friendlyValidityRange txValidityRange
+        ++  [ "withdrawals" .= friendlyWithdrawals txWithdrawals
+            | Just _ <- [withdrawalsSupportedInEra era]
+            ]
+    Left err -> Left $ displayError err
+  where
+    era = cardanoEra @era
+
+friendlyValidityRange
+  :: forall era
+  .  IsCardanoEra era
+  => (TxValidityLowerBound era, TxValidityUpperBound era) -> [(Text, Value)]
```

Comment:
This can be make more generic using `validityLowerBoundSupportedInEra` and `validityUpperBoundSupportedInEra`

Reply:
What do you mean?